source/common/network/connection_impl.cc

```diff
@@ -34,8 +35,9 @@ std::atomic<uint64_t> ConnectionImpl::next_global_id_;
 
 ConnectionImpl::ConnectionImpl(Event::DispatcherImpl& dispatcher, int fd,
```

Comment:
this is still not quite what I had in mind. Per the interface comment, here is what I would do:

Make the ConnectionImpl constructor look like this:
ConnectionImpl(Event::DispatcherImpl& dispatcher, int fd, const std::string& remote_address, const std::string& local_address);

This will force you to pre-populate the local address when creating the connection. For client connections, and redirect connections, just pass empty string for now, but in the future we can easily update to support all of those scenarios and also support UDS connections, etc.